src/Domain/Task/Handler/MarkAsDoneHandler.php

```diff
@@ -53,5 +57,58 @@ public function __invoke(MarkAsDone $command)
         if (!empty($contactName)) {
             $task->getAddress()->setContactName($contactName);
         }
+
+        $this->calculateCo2Impact($task);
+    }
+
+    private function calculateCo2Impact(Task $task) {
+
+        $taskList = $this->taskListRepository->findLastTaskListByTask($task);
+
+        if (!$taskList) {
+            $this->logger->error('Task was marked as finished but no corresponding tasklist was found'); // should not happen
+            return;
+        }
+
+        $coordinates = [];
+        $vehicle = $taskList->getVehicle();
+
+        if (!is_null($vehicle)) {
+            $coordinates[] = $taskList->getVehicle()->getWarehouse()->getAddress()->getGeo();
+        }
+
+        foreach ($taskList->getTasks() as $item) {
+            $coordinates[] = $item->getAddress()->getGeo();
+        }
+
+        // going back to the warehouse
+        if (!is_null($vehicle)) {
+            $coordinates[] = $taskList->getVehicle()->getWarehouse()->getAddress()->getGeo();
+        }
+
+        $route = $this->routing->route(...$coordinates)['routes'][0];
```

Comment:
We can live with it, but it's still a bit strange that we recalculate a route for each single task while we already calculated a route for the TaskList. I guess the issue is that we do not store route legs in the TaskList table?

Reply:
yes we just have the total distance stored in TaskList, we could maybe serialize the whole travel as JSON?

Reply:
I will add a comment for now